tensorflow/contrib/opt/python/training/elastic_average_optimizer.py

```diff
@@ -324,13 +393,69 @@ def _Add_sync_queues_and_barrier(enqueue_after_list):
       init_ops.append(state_ops.assign(lc_var, gc_var))
 
     init_op = control_flow_ops.group(*(init_ops))
+    if self._sync_flag == False:
+      return init_op
+
     sync_queue_op = _Add_sync_queues_and_barrier([init_op])
     return sync_queue_op
 
   def make_session_run_hook(self, is_chief, task_index):
     """Creates a hook to handle ElasticAverageOptimizerHook ops such as initialization."""
     return _ElasticAverageOptimizerHook(self, is_chief, task_index)
 
+  def swapping_saver(self, var_list=None, name='swapping_saver', **kwargs):
+    """Create a saver copy global_center_variable to trainable variables
+    Please call this function after all your variables created with EACustomGetter.
+    For evaluations or inference, use this saver during training.  It will save the
+    global_center_variable of the trained parameters under the original parameter names.
+    Args:
+      var_list: List of variables to save, as per `Saver()`.
+                If set to None, will save all the trainable_variables that have been
+                created before this call.
+      name: The name of the saver.
+      **kwargs: Keyword arguments of `Saver()`.
+    Returns:
+      A `tf.train.Saver` object.
+    Raises:
+      RuntimeError: global_center_variable is empty, please make sure
+                    this is called after model created and
+                    EACustomGetter is used when declaring you model
+    """
+    if not self._global_map:
+      raise RuntimeError('global_center_variable is empty, please make sure '
+                         'this is called after model created and '
+                         'ElasticAverageCustomGetter is used when declaring you model')
+
+    if var_list is None:
+      var_list = variables.trainable_variables()
+    if not isinstance(var_list, dict):
+      var_list = saver.BaseSaverBuilder.OpListToDict(var_list)
+
+    swapped_var_list = {}
+    has_global_step = False
```

Comment:
Why is it so important to care about the global step here? Should it be affected by the optimizer?

Reply:
In single GPU version, when saving a default model, global_step is saved.
When restoring model with the single GPU version code, global_step is needed in code.
And considering global_step is only 8byte, I added it to avoid potential user's problem when restoring model.

Reply:
Yes but I still do not understand why the global step has to be copied over in this case.

Reply:
OK, let me remove it, and user can still explicitly add it.